File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    length = len(stages)
    
    for i in range(1,N+1):
        cnt = stages.count(i)
        if length == 0:
            fail = 0
        else:
            fail = cnt/length
            
        answer.append((i,fail))
        length -= cnt
        
    answer = sorted(answer,key=lambda t:t[1], reverse=True)
    answer = [i[0] for i in answer]
    
    return answer

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] ch = new int[N+2];
        for(int i = 0; i<stages.length; i++){
            ch[stages[i]]++;
        }
        
        HashMap<Integer, Double> fail = new HashMap<>();
        double totalPerson = stages.length;
        
        for(int i=1; i<=N; i++){
            if(ch[i]==0){
                fail.put(i, 0.);
            }
            else{
            fail.put(i, (double)ch[i]/totalPerson);
            totalPerson -= ch[i];   
            }
        }
        
        List<Integer> keyList = new ArrayList<>(fail.keySet());
        Collections.sort(keyList, new Comparator<Integer>(){
            @Override
            public int compare(Integer o1, Integer o2){
                Double v1 = fail.get(o1);
                Double v2 = fail.get(o2);
                
                int compareResult = Double.compare(v2, v1);
                if(compareResult==0){
                    return Integer.compare(o1,o2);
                }
                return compareResult;
            }
        });
        
        int[] answer = new int[N];
        for(int i=0; i<keyList.size(); i++){
            answer[i] = keyList.get(i);
        }
        
        return answer;
    }
}